### PR TITLE
Conditionally clean up Gatekeeper Constraints

### PR DESCRIPTION
**Pull request by dhaiducek**
Listing Gatekeeper Constraints was refreshing the discovery API. This adds a global boolean to only attempt cleanup when a Constraint was previously synced successfully.

### CODE REVIEW

**Review by JustinKuli — DISMISSED** [review]
/hold for @mprahl 

LGTM! It's easier than I would have expected to transition the global back to false, so that's really nice.

File: controllers/templatesync/template_sync.go [hunk 1]
```diff
@@ -47,7 +47,10 @@ const (
 	ControllerName string = "policy-template-sync"
 )
 
-var log = ctrl.Log.WithName(ControllerName)
+var (
+	createdGkConstraint = false
```
Comment (mprahl) [inline controllers/templatesync/template_sync.go:51]: Could you make this a struct member of the reconciler instead?

File: controllers/templatesync/template_sync.go [hunk 2]
```diff
@@ -690,6 +695,7 @@ func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Requ
 			// Handle cluster scoped objects
 			if isClusterScoped {
 				addFinalizer = true
+				createdGkConstraint = true
```
Comment (mprahl) [inline controllers/templatesync/template_sync.go:698]: It'd be nice to not assume that any cluster scoped policy template is Gatekeeper.

File: controllers/templatesync/template_sync.go [hunk 3]
```diff
@@ -854,51 +866,29 @@ func (r *PolicyReconciler) cleanUpExcessTemplates(
...
 						Group:    utils.GvkConstraintTemplate.Group,
 						Resource: "constrainttemplates",
-						Version:  "v1beta1",
+						Version:  "v1",
 					},
 					namespaced: false,
 				})
+			} else {
+				// No ConstraintTemplates found -- reset boolean
+				*r.createdGkConstraint = false
```
Comment (mprahl) [inline controllers/templatesync/template_sync.go:888]: You'll want to explicitly set this to `false` on errors when the CRD doesn't exist error is encountered.

**Review by mprahl — APPROVED** [review]

**Review by mprahl — CHANGES_REQUESTED** [review]
Just one more comment.

File: controllers/templatesync/template_sync.go [hunk 4]
```diff
@@ -908,12 +898,45 @@ func (r *PolicyReconciler) cleanUpExcessTemplates(
...
+				}
+				// Log and ignore other errors to allow cleanup to continue since Gatekeeper may not be installed
+			} else {
+				reqLogger.Info(fmt.Sprintf("Ignoring ConstraintTemplate cleanup error: %s", err.Error()))
+				r.setCreatedGkConstraint(false)
+			}
 		} else {
 			reqLogger.Info(fmt.Sprintf("Ignoring ConstraintTemplate cleanup error: %s", err.Error()))
+			r.setCreatedGkConstraint(false)
 		}
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply controllers/templatesync/template_sync.go:939]: Oh yeah, that makes more sense.

**Pull request merged**